### ISSUE TITLE
Systemd loads emergency root shell after adding curlftpfs mountpoint into /etc/fstab

### ISSUE THREAD

**Issue by MikhailKasimov** [systemd/systemd#1053]
Hello! This report is dup of https://bugzilla.opensuse.org/show_bug.cgi?id=943146

In brief way -- to force systemd to mount curlftpfs correctly in pretty way on system start. The mentioned in openSUSE bz 943146 ways with systemd-fstab-generator or user-made .service don't work correctly instead of "old-ugly-..." unix-way by bash-scripting. Please, conside and help!

**Comment by michich** [comment]
Does it get mounted if you add "_netdev" to the mount options in fstab?
I also recommend adding "nofail" to avoid the switch to emergency mode in case the mount fails.

**Comment by MikhailKasimov** [comment]
Ok, the task has several variants. Before every 'sudo systemctl start home-k_mikhail-k_mikhail_mnts-yandex_ftp.mount' command 'sudo systemctl daemon-reload' is done.
1. We have /etc/systemd/system/home-k_mikhail-k_mikhail_mnts-yandex_ftp.mount:

```
[Unit]
After=network.target remote-fs.target local-fs.target

[Mount]
What=curlftpfs#ftp://ftp.yandex.ru/
Where=/home/k_mikhail/k_mikhail_mnts/yandex_ftp
Type=fuse
Options=rw,nosuid,nodev,relatime,user_id=1000,group_id=100
```

```
sudo systemctl start home-k_mikhail-k_mikhail_mnts-yandex_ftp.mount
```

Result:

```
home-k_mikhail-k_mikhail_mnts-yandex_ftp.mount - /home/k_mikhail/k_mikhail_mnts/yandex_ftp
   Loaded: loaded (/etc/systemd/system/home-k_mikhail-k_mikhail_mnts-yandex_ftp.mount; static)
   Active: failed (Result: exit-code) since Чтв 2015-08-27 18:05:26 EEST; 15s ago
    Where: /home/k_mikhail/k_mikhail_mnts/yandex_ftp
     What: curlftpfs#ftp://ftp.yandex.ru/
  Process: 6302 ExecMount=/bin/mount -n -t fuse -o rw,nosuid,nodev,relatime,user_id=1000,group_id=100 curlftpfs#ftp://ftp.yandex.ru/ /home/k_mikhail/k_mikhail_mnts/yandex_ftp (code=exited, status=1/FAILURE)

Авг 27 18:05:26 linux-mk500 mount[6302]: fuse: unknown option `relatime'
```

2.

```
[Unit]
After=network.target remote-fs.target local-fs.target

[Mount]
What=curlftpfs#ftp://ftp.yandex.ru/
Where=/home/k_mikhail/k_mikhail_mnts/yandex_ftp
Type=fuse
Options=rw,nosuid,_netdev
```

Result: yandex_ftp is displayed in file-manager as connected, but no objects can be listed.

3.

```
[Unit]
After=network.target remote-fs.target local-fs.target

[Mount]
What=curlftpfs#ftp://ftp.yandex.ru/
Where=/home/k_mikhail/k_mikhail_mnts/yandex_ftp
Type=fuse
Options=rw,nosuid,_netdev,user_id=1000
```

Result:

```
home-k_mikhail-k_mikhail_mnts-yandex_ftp.mount - /home/k_mikhail/k_mikhail_mnts/yandex_ftp
   Loaded: loaded (/etc/systemd/system/home-k_mikhail-k_mikhail_mnts-yandex_ftp.mount; static)
   Active: failed (Result: exit-code) since Чтв 2015-08-27 18:16:49 EEST; 2min 35s ago
    Where: /home/k_mikhail/k_mikhail_mnts/yandex_ftp
     What: curlftpfs#ftp://ftp.yandex.ru/
  Process: 6525 ExecMount=/bin/mount -n -t fuse -o rw,nosuid,_netdev,user_id=1000 curlftpfs#ftp://ftp.yandex.ru/ /home/k_mikhail/k_mikhail_mnts/yandex_ftp (code=exited, status=1/FAILURE)

Авг 27 18:16:49 linux-mk500 mount[6525]: fuse: unknown option `user_id=1000'
```
1. We temporary forget about our .mount-service and use the usual .service (yandex_ftp.service as I mentioned in OpenSUSE bz):

```
[Unit]
After=network.target

[Service]
Type=oneshot
ExecStart=/bin/bash -c 'curlftpfs ftp.yandex.ru \
/home/k_mikhail/k_mikhail_mnts/yandex_ftp'

[Install]
WantedBy=multi-user.target grafical.target
```

or, if we want to use ~/ instead of explicitly /home/k_mikhail, we need to add User=k_mikhail directive. Otherwise systemd conuts that ~/ is /root/.

Result: yandex_ftp is displayed in file-manager as connected, but no objects can be listed.

P.S. After some time I'll write about the results of _netdev and nofail in /etc/fstab. Generated by systemd-fstab-generator .mount:

Automatically generated by systemd-fstab-generator

```
[Unit]
SourcePath=/etc/fstab

[Mount]
What=curlftpfs#ftp://ftp.yandex.ru/
Where=/home/k_mikhail/k_mikhail_mnts/yandex_ftp
Type=fuse
Options=rw,nosuid,_netdev,nofail,relatime,user_id=1000,group_id=100
```

**Comment by poettering** [comment]
Your fuse file system cannot deal with `relatime` or `user_id=`, and it says so. I am not sure what this has to do with systemd.

if you don't want to block boot due to a file system, mark it as non-essential using the `nofail` fstab mount option.

If you want your file system to be mounted only after the network is up, use the `_netdev` fstab mount option. (But do note that your network management solution must be up for that).

Either way, I see only issues in your FUSE tool, and in the chosen configuration. Nothing to fix in systemd here. Closing.

**Comment by MikhailKasimov** [comment]
After rebooting with generated by systemd-fstab-generator .mount file result: yandex_ftp is displayed in file-manager as connected, but no objects can be listed (.mount file was with _netdev and nofail options). System mesage: "mount: only root can mount curlftpfs#ftp.yandex.ru on /home/k_mikhail/k_mikhail_mnts/yandex_ftp/".

Honestly (I don't try to offend anybody, but I'm very tired of searching the correct way to solve it...), the "ugly" traditional unix-way with simple .desktop, which contains Exec=/path/to/bash-script, where bash-script contains only three strings (1. /bin/bash 2. curlftpfs ftp.yandex.ru ~/k_mikhail_mnts and 3.exit 0), is much more convenient compared to "modern" systemd-way with its pile of .mount, .automount, .service and generators. But it was ineteresting to solve this simple task _only in systemd-way_.

2 Lennart: Please, pay attention, that simple service

[Unit]
After=network.target

[Service]
Type=oneshot
ExecStart=/bin/bash -c 'curlftpfs ftp.yandex.ru \
/home/k_mikhail/k_mikhail_mnts/yandex_ftp'

[Install]
WantedBy=multi-user.target grafical.target

should work in THE SAME way to bash-script I mentioned above (1. /bin/bash 2. curlftpfs ftp.yandex.ru ~/k_mikhail_mnts and 3.exit 0). Even with User=k_mikhail  directive (non-root user) in [Service] section. But bash-script connects ftp-server successfully in userspace and I can see the list of its folders, and systemd-service only shows me, that mountpoint is connected, but folder-list is unaccessable. As for me, this should be a hint to proceed the additional check systemd behavior. Seems to be the real bug. May be in parcer of bash-commands, defined in ExecStart=.

**Comment by michich** [comment]
The bash script that you run from a *.desktop file runs in your user session. That is not the case with script run from a systemd service.

You say that both ways result in a mount being done. Go ahead and compare the resulting mounts in each case using "findmnt". That should help point out any differences.

Doesn't the command fork? If it does, "Type=oneshot" is incorrect.

(As a sidenote, I find it remarkable that you consider *.desktop files to be the traditional unix-way :-) )

**Comment by MikhailKasimov** [comment]
Ok, will try to search further.

(.desktop creates XFCE, I only write simple and convenient bash-script, which does its job correctly (unix-way, yes). ;) )

**Comment by MikhailKasimov** [comment]
Resolved by creating mount-unit in /etc/system/system/ with such content:

```
[Unit]
After=network.target remote-fs.target local-fs.target

[Mount]
What=curlftpfs#ftp.yandex.ru
Where=/home/k_mikhail/k_mikhail_mnts/yandex_ftp
Type=fuse
Options=rw,nosuid,uid=1000,gid=100,allow_other

[Install]
WantedBy=multi-user.target grafical.target
```

After sudo systemctl daemon-reload && sudo systemctl start /etc/system/system/...mount && sudo systemctl enable /etc/system/system/...mount

everything works as needed. Thanks to everyone here and Ivan Shapovalov for help! :)

**Comment by RobinGhost** [comment]
@MikhailKasimov  I to have been trying to mount a remote folder using curlftpfs and systemd, like you were. I got the script to run manually, but when I try and run it through systemd it fails.

My script is

```
# this mounts a secure ftp remote path for local use
#! /bin/bash

curlftpfs -o ssl -o ssl_control -o no_verify_hostname -o no_verify_peer -d username:password@domain:port /mnt/ftpshare &
```

My systemd file is:

```
[Unit]
Description=ftp remote mount

[Service]
ExecStart=/home/ftpmount.sh
KillMode=process

[Install]
WantedBy=multi-user.target
```

When I run the service and check the status it comes back with:

```
● ftpmount.service - ftp remote mount
   Loaded: loaded (/etc/systemd/system/ftpmount.service; enabled)
   Active: failed (Result: exit-code) since Tue 2016-11-15 22:16:18 GMT; 23s ago
  Process: 1568 ExecStart=/home/ftpmount.sh (code=exited, status=203/EXEC)
 Main PID: 1568 (code=exited, status=203/EXEC)

Nov 15 22:16:18 openmediavault systemd[1]: Started ftp remote mount.
Nov 15 22:16:18 openmediavault systemd[1]: ftpmount.service: main process exited, code=exited, status=203/EXEC
Nov 15 22:16:18 openmediavault systemd[1]: Unit ftpmount.service entered failed state.
```

How would I convert my script to run in systemd like you managed to achieve?

**Comment by MikhailKasimov** [comment]
@scouseman  Try something simple, without having brain-rape with .sh->.service scheme, like :

```
[Unit]
After=network.target remote-fs.target

[Mount]
What=curlftpfs#ftp.domain
Where=your-mount-point-location
Type=fuse
Options=rw,nosuid,uid=<your_uid>,gid=<your_gid>,allow_other,ftp_port=21,user=user:password,<your_ssl_options>

[Install]
WantedBy=multi-user.target
```

Save this as `/etc/systemd/system/your-mount-point-location.mount` and do `systemctl start your-mount-point-location.mount`. And, if needed, do `systemctl enable your-mount-point-location.mount` to start your mount-point on system is being booted. And necessary read about your options on `-o` key in `man curlftpfs`. This should help!

**Comment by RobinGhost** [comment]
Thanks very much for your help @MikhailKasimov 
This is the basic mount script I used mnt.mount:

```
[Unit]
After=network.target remote-fs.target

[Mount]
What=curlftpfs#mydomain.com:2121
Where=/mnt/
Type=fuse
Options=rw,nosuid,uid=1000,gid=100,allow_other,user=username:password,-o ssl -o ssl_control -o no_verify_hostname -o no_verify_peer -d

[Install]
WantedBy=multi-user.target
```

I had to move the port out of option next to domain as it was defaulting to 21

I am getting the following error from status

```
● mnt.mount - /mnt
   Loaded: loaded (/etc/systemd/system/mnt.mount; disabled)
   Active: failed (Result: exit-code) since Wed 2016-11-16 20:47:37 GMT; 2s ago
    Where: /mnt
     What: curlftpfs#supplyanddemmand.tk:5757
  Process: 12634 ExecMount=/bin/mount -n curlftpfs#mydomain.com:2121 /mnt -t fuse -o rw,nosuid,uid=1000,gid=100,allow_other,user=username:password,-o ssl -o ssl_control -o no_verify_hostname -o no_verify_peer -d (code=exited, status=1/FAILURE)

Nov 16 20:47:37 openmediavault mount[12634]: fuse: unknown option `-o ssl -o ssl_control -o no_verify_hostname -o no_verify_peer -d'
Nov 16 20:47:37 openmediavault systemd[1]: mnt.mount mount process exited, code=exited status=1
Nov 16 20:47:37 openmediavault systemd[1]: Failed to mount /mnt.
Nov 16 20:47:37 openmediavault systemd[1]: Unit mnt.mount entered failed state.

```

This is the working curlftpfs command which I have working

`curlftpfs -o ssl -o ssl_control -o no_verify_hostname -o no_verify_peer -d username:password@domain:port /mnt/ftpshare`

Thanks again,

**Comment by arvidjaar** [comment]
> `After=network.target remote-fs.target`

Units that require network to be present on startup are usually expected to have

```
After=network-online.target
Wants=network-online.target
```

and of course whatever unit implements waiting for network to be online by your distribution should also be enabled. Additionally having `After=nss-lookup.target` will not hurt in this case also I am not sure whether this is actually used.

P.S. and BTW `remote-fs.target` is ordered _after_ remote mount units, at least those that are autogenerated from fstab, so line `After=remote-fs.target` is wrong here, it should be

```
After=remote-fs-pre.target
Before=remote-fs.target
```

**Comment by kajilama** [comment]
Thanks for a wonderful workaround @MikhailKasimov. 

I was having the same issue when attempting to mount my local ftp share (ftp built into router for USB storage) with my raspberry pi. I could manually mount the share with a direct curlftpfs command or manually mount all with "sudo mount -a" with the curlftpfs entry in fstab. But as soon as I rebooted raspberry pi, it basically failed to boot up, with basically nothing on the screen and in the logs. After attempting the same thing with a tiny debian distro, I found out that the fstab entry basically pushed the system into emergency mode locking out root. 

So after pending a couple of hours trying to fix this, I stumbled upon this page and attempted the workaround suggested above, and voila, it worked (though with a few minor adjustments). My raspberry pi rebooted and the ftp share is still mounted and accessible :)

Below is what worked for me, in my **/etc/systemd/system/mnt-nas.mount**:

`
[Unit]
After=network.target remote-fs.target

[Mount]
What=curlftpfs#ftp.domain/share-name
Where=/mnt/nas/
Type=fuse
Options=rw,uid=1000,gid=1000,allow_other,user=username:password,_netdev

[Install]
WantedBy=multi-user.target
`
